scipy/linalg/decomp_svd.py

```diff
@@ -279,6 +286,98 @@ def diagsvd(s, M, N):
         raise ValueError("Length of s must be M or N.")
 
 
+def higher_order_svd(
+    a: npt.ArrayLike, *, full_tensor: bool = True, check_finite: bool = True
+) -> Tuple[List[numpy.ndarray], numpy.ndarray]:
+    """Higher-order SVD (HOSVD)
+
+    Factorizes the M-D tensor `a` into a list of M matrices ``U_k`` containing
+    the left singular vectors of the unfolded tensor with respect to the
+    ``k``\\ th axis and a core tensor ``S`` such that
+    :math:`a = (U_1, U_2, \\dots, U_M) \\cdot S`.
+
+    Parameters
+    ----------
+    a : array_like
+        Tensor to decompose
+    full_tensor : bool, optional
+        If True (default), ``S`` is of the same shape as the input tensor `a`
+        and all ``U_k`` are of the shape ``(n, n)`` with ``n`` corresponding to
+        the ``k``\\ th axis of ``S``.
+        If False, the shapes of the ``U_k`` are ``(n, m)`` with ``n``
+        corresponding to the size of the ``k``\\ th axis of the input `a` and
+        ``m`` to the maybe reduced size of the ``k``\\ th axis of the core
+        tensor.
+    check_finite : bool, optional
+        Whether to check that the input matrix contains only finite numbers.
+        Disabling may give a performance gain, but may result in problems
+        (crashes, non-termination) if the inputs do contain infinities or NaNs.
+
+    Returns
+    -------
+    U : list of ndarray
+        List of the unitary matrices ``U_k`` with left singular vectors of the
+        tensor unfolding corresponding to the `k`\\ th axis.
+        Of shape ``(M, M)`` or ``(M, K)``, depending on `full_tensor`.
+    s : ndarray
+        The core tensor.
+        Of same or reduced shape compared to the input `a`, depending on
+        `full_tensor`.
+
+    Raises
+    ------
+    LinAlgError
+        If SVD computation does not converge.
+
+    .. versionadded:: 1.8.0
+
+    References
+    ----------
+    .. [1] Lieven De Lathauwer, Bart De Moor, and Joos Vandewalle (2000)
+           "A Multilinear Singular Value Decomposition"
+           SIAM Journal on Matrix Analysis and Applications,
+           21 (4), pp. 1253-1278. ISSN 1095-7162
+           https://doi.org/10.1137/S0895479896305696
+
+    Examples
+    --------
+    >>> from numpy.random import default_rng
+    >>> from scipy import linalg
+    >>> rng = default_rng()
+    >>> a = rng.random((2, 3, 4, 5))
+    >>> U, S = linalg.hosvd(a)
+    >>> S.shape
+    (2, 3, 4, 5)
+    >>> [i.shape for i in U]
+    [(2, 2), (3, 3), (4, 4), (5, 5)]
+
+    Reconstruct the original tensor:
+
+    >>> a1 = S
+    >>> for i in range(len(U)):
+    ...     a1 = np.tensordot(a1, U[i], (0, 1))
+    >>> np.allclose(a, a1)
+    True
+    """
+    a2 = _asarray_validated(a, check_finite=check_finite)
```

Comment:
Instead of having `a2`, I would just stay with `a` all along.

Reply:
The extra variable `a2` has been introduced due to the adding of typing information. mypy does allow the change of a variable type in the default config. `a2` is a ndarray but `a` is defined as ArrayLike only.

Reply:
I see... Then I guess it's fine otherwise we would need to do something like `a: numpy.ndarray = ...` to let MyPy know we are changing this. Or maybe there is another simpler/cleaner way.

Reply:
As far I understand this is a current limitation of mypy: https://mypy.readthedocs.io/en/stable/common_issues.html#redefinitions-with-incompatible-types

If I remember correctly the `a: numpy.ndarray = ...` does not fix this error message. mypy still has some drawbacks...

Reply:
ok thanks for the link and the information!

Reply:
Reassignment should be fine as long as it's a type narrowing, like going from `None | int` to `int`. 
So in this case it would be ok, as the broader much `npt.ArrayLike` is a supertype of `np.ndarray`.

As a counterexample, reassigning a `str` to a `float` would indeed cause a warning, as the latter 
is not a subtype of the former.

Reply:
Using `a: numpy.ndarray = _asarray_validated(a, check_finite=check_finite)` does not work. I just tried to do this but mypy outputs error message like:
 
```
scipy/linalg/decomp_svd.py:366: error: Name "a" already defined on line 289  [no-redef]
scipy/linalg/decomp_svd.py:371: error: Item "Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]]" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "ndim"  [union-attr]
```

Reply:
Can you use `a = _asarray_validated(a, check_finite=check_finite)` instead (without re-annotating `a`)?
In principle this should be enough.

Reply:
Exactly, `_asarray_validated` has no annotations yet. Since this is a little bit out of scope of this PR, what should I do?

Reply:
I personally have no objection adding some types to this single function here.

Reply:
If just adding a return-type to `_asarray_validated` doesn't cause any further errors, then it should be ok in my opninion.

Reply:
@BvB93 Could you pls check the new commits to add type hints for _asarray_validated? With this changes mypy is happy now

Reply:
A small nit, but otherwise it looks ok.
I'm honestly suprised we need to use `typing.cast` here; I think we're dealing with a mypy bug.
``` diff
toarray = cast(
-     Callable[..., np.ndarray],
+     Callable[["npt.ArrayLike"], np.ndarray],
    np.asarray_chkfinite if check_finite else np.asarray,
)
```

Reply:
The problem seems to be that numpy is not typed at this point yet: https://github.com/numpy/numpy/blob/main/numpy/lib/function_base.py#L426

Since the functions expects more arguments than just ArrayLike and we only care in scipy about the return type yet, I thought that the ellipsis would be ok for now?

Reply:
It's true that the function _can_ take more (optional) arguments, but they're not all that relevant here as they aren't used by `_asarray_validated`. We only care about the first positional argument.

Reply:
That is not true. The second last line of the functions uses the `dtype` keyword argument: `a = toarray(a, dtype=np.float_)`. Unfortunately, according to the doc `typing.Callable` does not support keyword args.

Reply:
> That is not true. The second last line of the functions uses the `dtype` keyword argument

You're right, let's stick to ellipsis for now. 

> Unfortunately, according to the doc `typing.Callable` does not support keyword args.

In principle it's possible to express these types of callables with [`__call__`-based protocol](https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols), but I'm not convinced it is worth the effort in this case (especially since it's a private function).